File: code/code22.py
from typing import List
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ans = []
        def backReturn(S,left,right):
            if (len(S) == 2 * n):
                ans.append("".join(S))
            if (left < n):
                S.append("(")
                backReturn(S,left+1,right)
                S.pop()
            if (right < left):
                S.append(")")
                backReturn(S,left,right+1)
                S.pop()
        backReturn([],0,0)
        return ans

File: code/code22.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        string str;
        generate(ans,n,str,0,0);
        return ans;
    }
    void generate(vector<string> &ans,int n,string &str,int left,int right) {
        if (str.size() == n) {
            ans.push_back(str);
        }
        if (left < n) {
            str.push_back('(');
            generate(ans,n,str,left+1,right);
            str.pop_back();
        }
        if (left > right) {
            str.push_back(')');
            generate(ans,n,str,left,right+1);
            str.pop_back();
        }
    }
};